### PR TITLE
Make CoreBundle optional

### PR DESCRIPTION
**Pull request by wbloszyk**
## Subject

This PR contains:
- SonataCoreBundle assets - 1 436 files.
- eventListerner for assets:install command - add sonatacore assets when CoreBundle is not installed
- drop SonataCoreBundle dependencies in favor for form and twig extensions 

I am targeting this branch, because this changes are BC.

Part of https://github.com/sonata-project/dev-kit/issues/697

## Changelog

```markdown
### Added
- Added `Sonata\AdminBundle\EventListener\AssetsInstallCommandListener` to add `sonatacore` asssets when SonataCoreBundle is not installed
### Changed
- ignore `sonata.admin.configuration.legacy_twig_text_extension` when SonataCoreBundle is not installed
### Removed
- remove all SonataCoreBundle dependencies
```

### CODE REVIEW

**Comment by wbloszyk** [comment]
Any idea why i can not use `%kernel.project_dir%`?? It should work.

**Comment by greg0ire** [comment]
Note that it only fails for `--prefer-lowest`. Try reproducing it in local. Then perform a dependency bisection: raise the constraint for half the dependency. Depending on the outcome, you will know which half is faulty. Rinse and repeat until you can narrow it down to just one dependency.

**Comment by wbloszyk** [comment]
@greg0ire 
You miss mu message?

https://github.com/sonata-project/SonataAdminBundle/pull/6096#issuecomment-639065174

**Comment by greg0ire** [comment]
Uuuuh no I did not miss your message… I was answering to it!

**Comment by wbloszyk** [comment]
@sonata-project/contributors RTM

**Comment by VincentLanglet** [comment]
Currently, the sonata core assets are used 

- multiple times in the standard layout
```
<script src="{{ asset('bundles/sonatacore/vendor/select2/select2_locale_' ~ localeForSelect2 ~ '.js') }}"></script>
```

- in the configuration
```
->defaultValue([
                                'bundles/sonatacore/vendor/bootstrap/dist/css/bootstrap.min.css',
                                'bundles/sonatacore/vendor/components-font-awesome/css/font-awesome.min.css',
                                'bundles/sonatacore/vendor/ionicons/css/ionicons.min.css',
                                'bundles/sonataadmin/vendor/admin-lte/dist/css/AdminLTE.min.css',
                                'bundles/sonataadmin/vendor/admin-lte/dist/css/skins/skin-black.min.css',
                                'bundles/sonataadmin/vendor/iCheck/skins/square/blue.css',

                                'bundles/sonatacore/vendor/eonasdan-bootstrap-datetimepicker/build/css/bootstrap-datetimepicker.min.css',

                                'bundles/sonataadmin/vendor/jqueryui/themes/base/jquery-ui.css',

                                'bundles/sonatacore/vendor/select2/select2.css',
                                'bundles/sonatacore/vendor/select2-bootstrap-css/select2-bootstrap.min.css',

                                'bundles/sonataadmin/vendor/x-editable/dist/bootstrap3-editable/css/bootstrap-editable.css',

                                'bundles/sonataadmin/css/styles.css',
                                'bundles/sonataadmin/css/layout.css',
                                'bundles/sonataadmin/css/tree.css',

                                'bundles/sonatacore/css/flashmessage.css',
                            ])
```

You didn't change this. How will it works if we don't require coreBundle ?

**Comment by wbloszyk** [comment]
@VincentLanglet 
I add event listener to `assets:install` command. This listener will copy this files to `sonatacore` directory simmilar like CoreBundle. 

I think better keep this `sonatacore` name. This will be easier to keep BC.

**Comment by VincentLanglet** [comment]
> I think better keep this `sonatacore` name. This will be easier to keep BC.

I agree

> I add event listener to assets:install command. This listener will copy this files to sonatacore directory simmilar like CoreBundle.

We should add a `NEXT_MAJOR` comment to explain what to do with this listener then.
Renaming the asset from sonatacore to sonataadmin ? 
Removing the listener ?

**Comment by wbloszyk** [comment]
> We should add a `NEXT_MAJOR` comment to explain what to do with this listener then.
> Renaming the asset from sonatacore to sonataadmin ?
> Removing the listener ?

We should not renaming this assets.
Listener will be remove when we drop Bootstrap 3 or use encore webpack insead.

I think we should not add any notes yet if we havent stable admin 4. What if we keep boostrap 3 support?

**Comment by VincentLanglet** [comment]
> Listener will be remove when we drop Bootstrap 3 or use encore webpack insead.
> 
> I think we should not add any notes yet if we havent stable admin 4. What if we keep boostrap 3 support?

Then you can add a note "Remove this listener when we drop boostrap 3" or "Remove this listener when we use webpack".

The idea is to help the next developers on this project :)

File: src/EventListener/AssetsInstallCommandListener.php [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+use Symfony\Component\Filesystem\Filesystem;
+use Symfony\Component\Finder\Finder;
+use Symfony\Component\HttpKernel\KernelInterface;
+
+/**
+ * This listener extends `assets:install` command when SonataCoreBundle will be not register. Files from `Resources/sonataCoreAssets`
+ * will be copy with the same result like SonataCoreBundle is register.
+ *
+ * This class should be remove when support for Bootstrap 3 will be ended or assets system will be remove in favor for encore webpack.
+ */
```
Comment (VincentLanglet) [inline src/EventListener/AssetsInstallCommandListener.php:33]: This helps a lot ! Thanks.

File: src/Resources/views/standard_layout.html.twig [hunk 2]
```diff
@@ -173,15 +173,15 @@ file that was distributed with this source code.
...
                                                                         {% endif %}
                                                                     </a>
                                                                 {% else %}
-                                                                    <span>{{ label|truncate(100) }}</span>
+                                                                    <span>{{ label|u.truncate(100) }}</span>
                                                                 {% endif %}
                                                             </li>
                                                         {% else %}
-                                                            <li class="active"><span>{{ label|truncate(100) }}</span></li>
+                                                            <li class="active"><span>{{ label|u.truncate(100) }}</span></li>
```
Comment (VincentLanglet) [inline src/Resources/views/standard_layout.html.twig:184]: Seems like there is other truncate to modify.
Cf: https://github.com/sonata-project/SonataAdminBundle/pull/6132/files

But you could wait for the PR to be merged.

BTW, is this change really CoreBundle related ?

Comment (wbloszyk) [reply src/Resources/views/standard_layout.html.twig:184]: Truncate was move to SonataTwigBundle 0.1.1. It can be merge.

Comment (VincentLanglet) [reply src/Resources/views/standard_layout.html.twig:184]: Can we revert this change then ? 
There is already another PR for the migration truncate => u.truncate.
Let's avoid potential conflict

Comment (wbloszyk) [reply src/Resources/views/standard_layout.html.twig:184]: I remove truncate commit.

File: src/EventListener/AssetsInstallCommandListener.php [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+        }
+
+        $io->newLine();
+
+        $copyUsed = false;
+        $exitCode = 0;
+        $validAssetDirs = [];
+
+        $bundle = $kernel->getBundle('SonataAdminBundle');
+        $originDir = $bundle->getPath().'/Resources/sonataCoreAssets';
```
Comment (core23) [inline src/EventListener/AssetsInstallCommandListener.php:119]: Is there a reason why this is called `sonataCoreAssets` and not `SonataCoreAssets` or just `SonataCoreBundle`?

File: src/EventListener/AssetsInstallCommandListener.php [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+        if (0 !== $exitCode) {
+            $io->error('Some errors occurred while installing assets.');
+        } else {
+            if ($copyUsed) {
+                $io->note('Some assets were installed via copy. If you make changes to these assets you have to run this command again.');
+            }
+
+            switch ($ioMethod) {
+                case 'success':$io->$ioMethod('All deprecated SonataCoreBundle assets from SonataAdminBundle were successfully installed.'); break;
+                case 'warning':$io->$ioMethod('All deprecated SonataCoreBundle assets from SonataAdminBundle were successfully installed.'); break;
```
Comment (core23) [inline src/EventListener/AssetsInstallCommandListener.php:165]:
```suggestion
                case 'success':
                case 'warning':$io->$ioMethod('All deprecated SonataCoreBundle assets from SonataAdminBundle were successfully installed.'); break;
```

File: src/EventListener/AssetsInstallCommandListener.php [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+use Symfony\Component\Finder\Finder;
+use Symfony\Component\HttpKernel\KernelInterface;
+
+/**
+ * This listener extends `assets:install` command when SonataCoreBundle will be not register. Files from `Resources/sonataCoreAssets`
+ * will be copy with the same result like SonataCoreBundle is register.
+ *
+ * This class should be remove when support for Bootstrap 3 will be ended or assets system will be remove in favor for encore webpack.
+ */
+class AssetsInstallCommandListener
```
Comment (core23) [inline src/EventListener/AssetsInstallCommandListener.php:34]: final

File: src/EventListener/AssetsInstallCommandListener.php [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Console\Style\SymfonyStyle;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symfony\Component\Filesystem\Exception\IOException;
+use Symfony\Component\Filesystem\Filesystem;
+use Symfony\Component\Finder\Finder;
+use Symfony\Component\HttpKernel\KernelInterface;
+
+/**
+ * This listener extends `assets:install` command when SonataCoreBundle will be not register. Files from `Resources/sonataCoreAssets`
```
Comment (core23) [inline src/EventListener/AssetsInstallCommandListener.php:29]: NEXT_MAJOR missing

File: src/Object/MetadataInterface.php [hunk 7]
```diff
@@ -13,29 +13,55 @@
 
 namespace Sonata\AdminBundle\Object;
 
-/**
- * NEXT_MAJOR: Remove CoreBundle dependency.
- */
-interface MetadataInterface extends \Sonata\CoreBundle\Model\MetadataInterface
-{
-    public function getTitle(): string;
+if (interface_exists('Sonata\CoreBundle\Model\MetadataInterface')) {
```
Comment (core23) [inline src/Object/MetadataInterface.php:16]: Please use simple proxies:
https://github.com/sonata-project/SonataMediaBundle/pull/1744#discussion_r437261570

File: src/EventListener/AssetsInstallCommandListener.php (hunk 3 above)
Comment (wbloszyk) [reply src/EventListener/AssetsInstallCommandListener.php:119]: I just name it like this. SonataCoreBundle will be fonfusing. I will change it to SonataCoreBundleAssets when i finish clearing my computer from dust.

Comment (wbloszyk) [reply src/EventListener/AssetsInstallCommandListener.php:119]: I just name it like this. SonataCoreBundle will be fonfusing. I will change it to SonataCoreBundleAssets when i finish clearing my computer from dust.

File: src/EventListener/AssetsInstallCommandListener.php (hunk 6 above)
Comment (wbloszyk) [reply src/EventListener/AssetsInstallCommandListener.php:29]: We can't add NEXT_MAJOR until we don't know what we do with assets in next Sonata 4. I think we can remove it just in Sonata 5.

File: src/EventListener/AssetsInstallCommandListener.php [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+use Symfony\Component\Filesystem\Exception\IOException;
+use Symfony\Component\Filesystem\Filesystem;
+use Symfony\Component\Finder\Finder;
+use Symfony\Component\HttpKernel\KernelInterface;
+
+/**
+ * This listener extends `assets:install` command when SonataCoreBundle will be not register. Files from `Resources/private/SonataCoreBundleAssets`
+ * will be copy with the same result like SonataCoreBundle is register.
+ *
+ * This class should be remove when support for Bootstrap 3 will be ended or assets system will be remove in favor for encore webpack.
```
Comment (jordisala1991) [inline src/EventListener/AssetsInstallCommandListener.php:32]: This class should be removed on Sonata 4, all assets belong here. On Sonata 4 we will decide if we remove bootstrap 3 or what, but it is not dependent, right?

File: src/Export/Exporter.php [hunk 9]
```diff
@@ -13,19 +13,26 @@
...
-);
-
-/**
- * NEXT_MAJOR: remove this class.
- *
- * @deprecated since sonata-project/admin-bundle 3.14, to be removed in 4.0.
- */
-class Exporter extends BaseExporter
-{
+if (class_exists('Sonata\CoreBundle\Exporter\Exporter')) {
```
Comment (jordisala1991) [inline src/Export/Exporter.php:16]: BaseExporter::class

File: src/Object/MetadataInterface.php [hunk 10]
```diff
@@ -13,10 +13,21 @@
...
+use Sonata\BlockBundle\Meta\MetadataInterface as BlockMetadataInterface;
+use Sonata\CoreBundle\Model\MetadataInterface as CoreMetadataInterface;
+
+// NEXT_MAJOR: Remove CoreBundle dependency.
+if (interface_exists(CoreMetadataInterface::class)) {
+    interface InternalMetadataInterface extends CoreMetadataInterface, BlockMetadataInterface
+    {
+    }
+} else {
+    interface InternalMetadataInterface extends BlockMetadataInterface
```
Comment (jordisala1991) [inline src/Object/MetadataInterface.php:25]: Same metadata, does it have to do with BlockBundle, can we remove this extends?

File: src/Export/Exporter.php (hunk 9 above)
Comment (wbloszyk) [reply src/Export/Exporter.php:16]: CoreExporter for this.
BaseExporter for sonata/exporter

File: src/Object/MetadataInterface.php (hunk 10 above)
Comment (wbloszyk) [reply src/Object/MetadataInterface.php:25]: https://github.com/sonata-project/SonataMediaBundle/pull/1744#discussion_r437643302

File: src/EventListener/AssetsInstallCommandListener.php (hunk 8 above)
Comment (wbloszyk) [reply src/EventListener/AssetsInstallCommandListener.php:32]: https://github.com/sonata-project/SonataAdminBundle/pull/6096#discussion_r437312790

This `sonatacore` assets are needed in whole sonata 3 (bootstrap 3). Sonata have layouts in 2 places `AminBundle` and `PageBundle` (this require AdminBundle). This mean move this assets to AdminBundle is mandatory.

Existing projects using this assets. We should allow use it in Sonata 4 where should be two assets system - deprecated assets for keep BC and new webpack. 

This class(and sonatacore assets) should be remove in Sonata 5 then or we must add webpack in Sonata 3.

File: src/DependencyInjection/SonataAdminExtension.php [hunk 11]
```diff
@@ -324,12 +325,19 @@ private function replacePropertyAccessor(ContainerBuilder $container): void
 
     private function configureTwigTextExtension(ContainerBuilder $container, XmlFileLoader $loader, array $config): void
     {
+        $bundles = $container->getParameter('kernel.bundles');
+
         $container->setParameter('sonata.admin.configuration.legacy_twig_text_extension', $config['options']['legacy_twig_text_extension']);
         $loader->load('twig_string.xml');
 
-        if (false !== $config['options']['legacy_twig_text_extension']) {
+        if (isset($bundles['SonataCoreBundle']) && false !== $config['options']['legacy_twig_text_extension']) {
```
Comment (wbloszyk) [inline src/DependencyInjection/SonataAdminExtension.php:333]: @franmomu 
You don't have to:
https://github.com/sonata-project/SonataAdminBundle/pull/6132#issuecomment-642455452

**Comment by wbloszyk** [comment]
@greg0ire @core23 Can you recheck it? Will be nice to have more approvals here.

**Comment by wbloszyk** [comment]
Code is copy from assests:install command. Only Paths and message are changed. I tested it on sandbox.

**Comment by jordisala1991** [comment]
Thank you @wbloszyk

File: src/EventListener/AssetsInstallCommandListener.php [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+    }
+
+    public function copySonataCoreBundleAssets(ConsoleTerminateEvent $event)
+    {
+        $command = $event->getCommand();
+        $application = $command->getApplication();
+
+        try {
+            $coreBundle = $application->getKernel()->getBundle('SonataCoreBundle');
+        } catch (\Exception $e) {
```
Comment (phansys) [inline src/EventListener/AssetsInstallCommandListener.php:62]: Shouldn't we catch `\Throwable`?

File: src/EventListener/AssetsInstallCommandListener.php [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+            if (self::METHOD_COPY === $method) {
+                $copyUsed = true;
+            }
+
+            if ($method === $expectedMethod) {
+                $ioMethod = 'success';
+            } else {
+                $ioMethod = 'warning';
+            }
+        } catch (\Exception $e) {
```
Comment (phansys) [inline src/EventListener/AssetsInstallCommandListener.php:151]: Shouldn't we catch `\Throwable`?

File: src/EventListener/AssetsInstallCommandListener.php (hunk 12 above)
Comment (wbloszyk) [reply src/EventListener/AssetsInstallCommandListener.php:62]: Here mybe, but it work.

Comment (wbloszyk) [reply src/EventListener/AssetsInstallCommandListener.php:62]: Here mybe, but it work.

File: src/EventListener/AssetsInstallCommandListener.php (hunk 13 above)
Comment (wbloszyk) [reply src/EventListener/AssetsInstallCommandListener.php:151]: Here not. It is copy from symfony assets:install command.

File: src/EventListener/AssetsInstallCommandListener.php (hunk 12 above)
Comment (wbloszyk) [reply src/EventListener/AssetsInstallCommandListener.php:62]: InvalidArgumentException -> LogicException -> Exception
```php
    /**
     * Returns a bundle.
     *
     * @param string $name Bundle name
     *
     * @return BundleInterface A BundleInterface instance
     *
     * @throws \InvalidArgumentException when the bundle is not enabled
     */
    public function getBundle($name);
```

**Pull request merged**